### ISSUE TITLE
Split keys over named batch axis

### ISSUE THREAD

**Issue by dieterichlawson** [google/jax#7708]
A pattern that I find myself using often in jax is defining a `loss`, and then defining an outer `average_loss` that `vmap`s the inner `loss` function and then averages, i.e.
```
def loss(key, params, x):
  return ...

def average_loss(key, params, xs):
  keys = jax.random.split(key, num=xs.shape[0])
  losses =  vmap(loss, in_axes=(0, None, 0))(keys, params, xs)
  return jnp.mean(losses)
```
I can use `fold_in` and a named axis to make this cleaner:
```
def loss(key, params, x):
  key = random.fold_in(key, lax.axis_index('batch'))
  ...
  return jnp.mean(loss, 'batch')
  
average_loss = vmap(loss, in_axes=(None, None, 0), axis_name='batch')
```
But these two approaches don't give the exact same results. It would be nice to have an 'official' way to split keys over batch axes.

**Comment by froystig** [comment]
Even if we don't dictate an official recipe, we should see if we can make these behave equivalently.

**Comment by apaszke** [comment]
It seems like this is because `split` folds twice as many indices as `fold_in`, but it's not explained anywhere. It might have to do with the design of threefry. Any clues @mattjj @hawkinsp
